File: a013. 羅馬數字/C/Lyu.c
#include <stdio.h>
#include <stdlib.h>

int decodeRoma(char chr, int *pos);
int encodeRoma(int number);

int main(void)
{
    char chr;
    int num[2] = {0, 0}, indexOfNum = 0;
    int current_pos = 7, previous_pos = 7;
    int decode_num[7] = {1, 5, 10, 50, 100, 500, 1000};
    scanf("%c", &chr);
    while (1)
    {
        if (chr == '#') // end loop
            break;
        else if (chr == 10) // '\0'
        {
            int result = num[0] - num[1];
            if (result < 0)
                result = -result;
            encodeRoma(result);
            num[0] = num[1] = indexOfNum = 0;
            current_pos = previous_pos = 7;
        }
        else if (chr == ' ')
        {
            indexOfNum++;
            current_pos = previous_pos = 7;
        }
        else
        {
            num[indexOfNum] += decodeRoma(chr, &current_pos);
            if (current_pos > previous_pos)
                num[indexOfNum] -= 2 * decode_num[previous_pos];
            previous_pos = current_pos;
        }
        scanf("%c", &chr);
    }
}

int decodeRoma(char chr, int *pos)
{
    char code[7] = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};
    int number = 1;
    for (int i = 0; i < 7; i++)
    {
        if (chr == code[i])
        {
            *pos = i;
            return number;
        }
        number *= (i % 2 == 0) ? 5 : 2;
    }
    printf("%c is not in decode list\n", chr);
    return -1;
}

int encodeRoma(int number)
{
    char code[7] = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};
    int code_ptr[4] = {6, 4, 2, 0};
    int digit = 1000, current_num = 0;
    for (int i = 0; i < 4; i++)
    {
        current_num = number / digit % 10;
        if (current_num <= 3)
        {
            for (int j = 0; j < current_num; j++)
                printf("%c", code[code_ptr[i]]);
        }
        else if (current_num == 4)
        {
            printf("%c%c", code[code_ptr[i]], code[code_ptr[i] + 1]);
        }
        else if (current_num <= 8)
        {
            printf("%c", code[code_ptr[i] + 1]);
            for (int j = 0; j < current_num - 5; j++)
                printf("%c", code[code_ptr[i]]);
        }
        else if (current_num == 9)
        {
            printf("%c%c", code[code_ptr[i]], code[code_ptr[i] + 2]);
        }
        digit /= 10;
    }
    if (number == 0)
        printf("ZERO");
    printf("\n");
    return 0;
}

File: a013. 羅馬數字/Java/Lyu.java
import java.util.Scanner;

public class Lyu {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        while (scan.hasNext()) {
            String line = scan.nextLine();
            if (line.charAt(0) == '#')
                break;
            String[] strings = (line.split(" "));
            int[] nums = new int[2];
            nums[0] = decodeString(strings[0]);
            nums[1] = decodeString(strings[1]);
            printResult(nums);
        }
        scan.close();
    }

    private static int decodeRoma(char chr) {
        char[] romaNums = { 'I', 'V', 'X', 'L', 'C', 'D', 'M' };
        int num = 1;
        for (int i = 0; i < romaNums.length; i++) {
            if (chr == romaNums[i])
                return num;
            else {
                num *= (i % 2 == 0) ? 5 : 2;
            }
        }
        return 0;
    }

    private static int decodeString(String str) {
        int result = 0, decodeNum = 0;
        int previousNum = 1001;
        for (char c : str.toCharArray()) {
            decodeNum = decodeRoma(c);
            result += decodeNum;
            if (previousNum < decodeNum) {
                result -= 2 * previousNum;
            }
            previousNum = decodeNum;
        }
        return result;
    }

    private static void printResult(int[] nums) {
        char[] romaNums = { 'I', 'V', 'X', 'L', 'C', 'D', 'M' };
        int index = 6, digit = 1000, digitNum = 0;
        int result = (nums[0] > nums[1]) ? nums[0] - nums[1] : nums[1] - nums[0];
        if (result == 0) {
            System.out.println("ZERO");
            return;
        }
        while (digit > 0) {
            digitNum = result / digit;
            if (digitNum < 4) {
                for (int i = 0; i < digitNum; i++) {
                    System.out.print(romaNums[index]);
                }
            } else if (digitNum == 4) {
                System.out.printf("%c%c", romaNums[index], romaNums[index + 1]);
            } else if (digitNum < 9) {
                System.out.print(romaNums[index + 1]);
                for (int i = 5; i < digitNum; i++) {
                    System.out.print(romaNums[index]);
                }
            } else {
                System.out.printf("%c%c", romaNums[index], romaNums[index + 2]);
            }
            result -= digitNum * digit;
            digit /= 10;
            index -= 2;
        }
        System.out.println();
    }
}